### DISCUSSION TITLE
Amber Electric dynamic tariffs, general chat / observations

### DISCUSSION THREAD

**Discussion by rob0101** [netzero-labs/netzero#12] [General]
If there's a better place to discuss the Amber Electric dynamic tariffs with both zigam and other customers, please point me there.

I've played with this feature today.  Thanks zigam so much for adding it.

It coped well with SA's rollercoaster pricing on the RELE2W tariff during the day.  I'm keen to see how it works overnight when I know I'll need some grid power to navigate the morning peak, but that's more a powerwall algorithm test than Netzero specific. It'll be interesting to see if it adapts quickly.

I'm seeing the tariff data update every 5 minutes, which is great, and definitely needed if we're going to use it as a smartshift alternative.

One snag - I'm seeing estimate prices rather than the actual prices being used.  This is an annoying quirk of Amber's API, they release an estimate price soon after each 5m on the clock, and replace it with one based on an actual price anytime from 30s or 90s (or more) later.   Amber need to improve this, Localvolts do a better job of this I'm led to believe.    Most of the time estimate =~ actual, but it's not uncommon to get very different prices.

Amber should look into tariff based forecasts rather than customer forecasts. At some stage lots of people will be triggering the fetching of the exact same forecasts from Amber's API. There are many dozens of tariffs supported, but most customers are on a much smaller subset of those tariffs, so lots of duplicated effort.

![IMG_DB23179F94F0-1](https://github.com/netzero-labs/netzero/assets/19808721/c1a9d798-12b4-436c-bf1d-7db77b054698)

**Comment by zigam** [comment]
Thanks for the feedback, @rob0101!

> One snag - I'm seeing estimate prices rather than the actual prices being used. This is an annoying quirk of Amber's API, they release an estimate price soon after each 5m on the clock, and replace it with one based on an actual price anytime from 30s or 90s (or more) later. Amber need to improve this, Localvolts do a better job of this I'm led to believe. Most of the time estimate =~ actual, but it's not uncommon to get very different prices.

I'll have a look at that, it might be that I just need to use the current price API in addition to the forecasts, and overwrite the current time period.

@purcell-lab explained the actual price a bit differently, he said it's only locked in in the last 5 minutes of a 30 minute period.  Either way, this seems like a lot of complexity that Amber passes down to customers; I understand the desire to track wholesale prices closely, but I wonder if there are diminishing returns here.  The other real-time plans I'm adding support for (Octopus Energy, Tibber) announce tariffs 24h in advance and don't change; this gives more time to adapt behavior, both for humans and batteries.

**Reply by rob0101** [reply]
Prices change twice every 5 minutes with Amber. 
A few seconds after the 5m interval they publish a new estimated 5m price that is also used to compute a new 30m estimate. They then wait for the market to release the actual 5m price. 

This waiting can only be done by the customer with a wait then re-poll, using up api limits. I find the wait needed to be anywhere from 30s to 90s most of the time.  Occasionally it can fail to update at all, but it's rare. 

Amber's lead developer said ages ago that they'd offer a push based solution, but as with all changes from them it never happens quickly. Polling in this model sucks. 

For netzero, one option might be to postpone the api call by 30+ seconds to get a better price to work from. Mark or others might have different theories.

**Reply by zigam** [reply]
Update here: I looked at the code and it is using CurrentInterval, so no changes needed there.  It's just that the timing of the request matters as @rob0101 explained.  Netzero currently updates tariffs every 5 minutes, but independently of clock time.  I'll change that to update tariffs at minute 2,7,12,17,22,... which should increase the chances of an ActualInterval instead of ForecastInterval.

**Comment by purcell-lab** [comment]
> I'll have a look at that, it might be that I just need to use the current price API in addition to the forecasts, and overwrite the current time period.

Yes,  This is a good catch by @rob0101, which I had overlooked in our discussions.

To create the forecast matrix for the next 24 hours you should include the current price API as the now entity and then the forecasts for the subsequent periods. The current price endpoint also continues to change during the 30 minute time periods, so you should continue to inject that every 5 minutes as well as the forecasts.

This is how I am calculating the matrix in Home Assistant:

```yaml
        "prod_price_forecast": {{
          ([states('sensor.amber_feed_in_price')|float(0)] +
          (state_attr('sensor.amber_feed_in_forecast', 'forecasts')|map(attribute='per_kwh')|list))
          | tojson 
        }},
        "load_cost_forecast": {{
          ([states('sensor.amber_general_price')|float(0)] + 
          state_attr('sensor.amber_general_forecast', 'forecasts') |map(attribute='per_kwh')|list) 
          | tojson 
        }}
```

```
"prod_price_forecast": [0.13, 0.17, 0.05, 0.05, 0.05, 0.04, 0.04, 0.04, 0.05, 0.05, 0.06, 0.06, 0.06, 0.06, 0.07, 0.08, 0.14, 0.11, 0.1, 0.13, 0.15, 0.2, 0.19, 0.16, 0.14, 0.15, 0.15, 0.15, 0.14, 0.12, 0.18, 0.15, 0.15, 0.15, 0.15, 0.2, 0.2, 0.12, 0.11, 0.11, 0.09, 0.09],
"load_cost_forecast": [0.22, 0.26, 0.14, 0.14, 0.12, 0.12, 0.12, 0.12, 0.13, 0.13, 0.13, 0.14, 0.14, 0.14, 0.15, 0.16, 0.23, 0.19, 0.33, 0.36, 0.38, 0.44, 0.42, 0.4, 0.37, 0.38, 0.39, 0.39, 0.24, 0.21, 0.27, 0.25, 0.25, 0.24, 0.25, 0.3, 0.3, 0.21, 0.2, 0.2, 0.18, 0.18]
```

![image](https://github.com/netzero-labs/netzero/assets/79175134/609bfa92-28ca-45f6-a490-2921d8923e3a)

![Screenshot_20240504-070424](https://github.com/netzero-labs/netzero/assets/79175134/397ffc9a-0b79-47ef-8c82-45bc037c2eec)

So the current buy price should be 0.22 $/kWh, not 0.23 as currently shown in Tesla app.

**Reply by rob0101** [reply]
If practical, wait for the actual current price, not the short lived estimated current price.

For all my price based actions, I ignore the estimated prices as they can easily be an order of magnitude wrong.
Right now in SA the 11:30 spot estimate is 11c and the actual became 4c.

**Comment by purcell-lab** [comment]
Overall, I am finding the netzero injected prices combined with autonomous mode are producing very good results and a lot smoother than SmartShift (I haven't been though a price spike which will be the true test).

Last night autonomous mode exported solidly exactly when I would have liked to it to from 5pm- 6pm when the feed in price was highest for the day, then leaving it in autonomous mode it ran the house solidly from battery for the rest of the night:

![image](https://github.com/netzero-labs/netzero/assets/79175134/ed1b2561-123a-4ffa-b999-b8766f429872)

Some nights it has topped the battery up overnight, when there was low production and high prices forecast for the following day, other days it has preserved the battery at a mid SOC.  It has also done a great job in charging the battery during the day from the grid at the cheapest prices of the day.  It is also working well with the Tesla Charge on Solar feature for my vehicles and it seems to preserve charge correctly when the prices are low and runs heavy loads from the grid, rather than consuming the battery during low prices.

![image](https://github.com/netzero-labs/netzero/assets/79175134/d814899a-627e-45e7-b234-30ea6b742358)

As I mentioned at the start the true test will be how it performs for a price spike, but I am optimistic given it has been exporting strongly with relatively low evening/ morning peaks.

**Reply by rob0101** [reply]
I turned mine off last night (first night) when TBC started buying around 9:15 but with prices expected to be cheaper later. 

My own model called for a "do nothing yet" response at that moment. It was an expensive night so didn't want to experiment with high prices with nobody watching.

Might have another look over the next few nights for a good time to trust it.

During the day it did what I wanted and charged at different rates, which is something smartshift doesn't do (it's 0 or the 5kW max).

**Reply by purcell-lab** [reply]
My charging cycle was very strong today, it waited until the cheapest prices of the day and then started importing to charge the battery, but then when the price went up a couple of cents the charging rate slowed and then stopped and started again when the price dropped again. 

![Screenshot_20240504-133439](https://github.com/netzero-labs/netzero/assets/79175134/4253e105-f60d-4e07-a471-a5d6d3010001)

**Comment by purcell-lab** [comment]
Worked well last night this morning. 

Held back some charge for overnight and then charged up a little overnight to fully discharge for the morning. 

![Screenshot_20240520-085024](https://github.com/netzero-labs/netzero/assets/79175134/a4aebcd8-a5ce-45c6-ac18-db70a5ac3aec)
![Screenshot_20240520-085011](https://github.com/netzero-labs/netzero/assets/79175134/14f9dd8d-f93b-4758-a475-1ee8bdc62519)
